Add `preprocess_transforms` and `setup_execution_config` (#6617)
PennyLaneAI/pennylane: doc/development/plugins.rst

Preprocessing
-------------

> There are two components of preprocessing circuits for device execution:

Reviewer: Mind updating Line 123 below:

```
These two tasks can be extracted into private methods or helper functions if that improves source
code organization.
```

```
These two tasks equate to ``preprocess_transforms`` and ``setup_execution_config``.
```

They don't need to be private anymore.

Author: True...

.. code-block:: python

    execution_config = dev.setup_execution_config(initial_config)
    transform_program = dev.preprocess_transforms(execution_config)
    batch, fn = transform_program(initial_batch)
    fn(dev.execute(batch, execution_config))

PennyLane can potentially provide a default implementation of a transform program through :meth:`~.devices.Device.preprocess_transforms`,
which should be sufficient for most plugin devices. This requires that a TOML-formatted configuration
file is defined for your device. The details of this configuration file is described :ref:`the next section <device_capabilities>`.
The default preprocessing program will be constructed based on what is declared in this file if provided.

> Note that the you should still override :meth:`~.devices.Device.setup_execution_config`.

Reviewer 2: ```suggestion
Note that you should still override the :meth:`~.devices.Device.setup_execution_config`.
```

> Note that the you should still override :meth:`~.devices.Device.setup_execution_config`.

Reviewer 3: Is this true? Or is it more "you way still want to"? If you are working from a TOML file and default, are you not using the default `preprocess_transforms` and the `DefaultExecutionConfig`, so that you don't need to overwrite either?

Reviewer 3: And should we comment more here on what the execution config is and when you would need to update it on the device instead of just using the default? Our initial description of `preprocess` really focused on the `TransformProgram` part, with a few notes at the end about the `ExecutionConfig`.

Author: I guess I didn't phrase it well enough. What I'm trying to get at is that there is a default preprocessing program which is more than an empty `TransformProgram`, but there is no default processing of the execution config, and all you're getting is the initial default config object. Although following our meeting this morning, we are adding a default `setup_execution_config` afterall, and I'll change the wording here to reflect that. To your second point, there is an entire section below talking about what execution config is.

Reviewer 3: Wrt the first point: sounds good!

Regarding the second point, the "Preprocess" section starts by saying, "There are two components of preprocessing circuits for device execution: Create a TransformProgram and Setup the ExecutionConfig.” While there's a link at the end of the functions list about the ExecutionConfig, as a user, I’d probably skim the section and expect both concepts to be covered - and I would stop when I hit the list of functions at the end, so I would never see the final line that links to the ExecutionConfig section. It can be frustrating if the intro sets up two ideas but only explains one, leaving the other unaddressed.

To make it clearer, I suggest moving the ExecutionConfig section link up to the end of the intro, and adding something like, "For more details on ExecutionConfig, see here—the rest of this section will focus on preprocess implementation." This way, the section clearly delivers on its promise and avoids confusion.

Author: Done

.. code-block:: python

    execution_config = dev.setup_execution_config(initial_config)
    transform_program = dev.preprocess_transforms(execution_config)
    batch, fn = transform_program(initial_batch)
    fn(dev.execute(batch, execution_config))

This section will focus on step 1, see the section on the :ref:`**Execution Config** <execution_config>`

> below for more information on step 2.

Reviewer: ```suggestion
below for more information on ``setup_execution_config``.
```

PennyLane can potentially provide a default implementation of a transform program through :meth:`~.devices.Device.preprocess_transforms`,
which should be sufficient for most plugin devices. This requires that a TOML-formatted configuration
file is defined for your device. The details of this configuration file is described :ref:`the next section <device_capabilities>`.
The default preprocessing program will be constructed based on what is declared in this file if provided.

> Alternatively, you could override the :meth:`~.devices.Device.preprocess_transforms` method with a
> completely customized implementation.

Reviewer: ```suggestion
You could override the :meth:`~.devices.Device.preprocess_transforms` method with a
completely customized implementation, or extend the default behaviour by adding new transforms.
```
